RHTAP-4843: Dynamic Dependency Resolution

otaviof opened the pull request:
# Dynamic Dependency Resolution

The dynamic dependency resolution enables the installer to automatically determine the correct deployment order and target namespaces for Helm charts based on annotations within their respective `Chart.yaml` files.

It works by:

- `product-name` annotation: Links a Helm chart to a specific product defined in `config.yaml`'s `.tssc.products[]`, which dictates its namespace.
- `depends-on` annotation: Declares direct dependencies, ensuring charts are deployed only after their prerequisites are met.

The resolution logic involves two phases:

1. *Resolving Enabled Products*: It first identifies and orders charts associated with enabled products, recursively including their dependencies.
2. *Resolving Remaining Dependencies*: It then processes any standalone dependency charts not directly tied to a product, ensuring they are placed correctly in the deployment sequence.

## Dependency Namespaces

Namespaces are determined by the product-name annotation: product-associated charts use the product's configured namespace `.tssc.products[].namespace`, while common dependencies use the installer's default namespace. This ensures a predictable and correct deployment flow.

As well the `config.yaml` has been adapted to have a `.tssc.products[]` slice instead of a map, removing the hardcoded product names from the code base.

# Topology Subcommand

This pull-request introduces a new topology subcommand that allows users to inspect the installer's dependency graph. It resolves and displays the deployment order of Helm charts, their target namespaces, associated products, and declared dependencies, providing a clear overview of the installation topology.

```
$ bin/tssc topology
Index  Chart                  Namespace  Product                    Depends-On
 1     tssc-openshift         tssc
 2     tssc-subscriptions     tssc
 3     tssc-infrastructure    tssc                                  tssc-openshift, tssc-subscriptions
 4     tssc-backing-services  tssc       Keycloak                   tssc-openshift, tssc-subscriptions, tssc-infrastructure
 5     tssc-tpa-realm         tssc                                  tssc-openshift, tssc-subscriptions, tssc-backing-services
 6     tssc-tpa               tssc-tpa   Trusted Profile Analyzer   tssc-openshift, tssc-subscriptions, tssc-infrastructure, tssc-backing-services, tssc-tpa-realm
 7     tssc-tas               tssc-tas   Trusted Artifact Signer    tssc-openshift, tssc-subscriptions, tssc-infrastructure, tssc-backing-services
 8     tssc-pipelines         tssc       OpenShift Pipelines        tssc-openshift, tssc-subscriptions, tssc-tas
 9     tssc-gitops            tssc       OpenShift GitOps           tssc-openshift, tssc-subscriptions
10     tssc-app-namespaces    tssc                                  tssc-pipelines, tssc-tpa
11     tssc-dh                tssc-dh    Developer Hub              tssc-openshift, tssc-subscriptions, tssc-infrastructure, tssc-pipelines, tssc-gitops, tssc-tas, tssc-tpa, tssc-app-namespaces
12     tssc-acs               tssc-acs   Advanced Cluster Security  tssc-openshift, tssc-subscriptions
13     tssc-acs-test          tssc                                  tssc-acs
14     tssc-integrations      tssc                                  tssc-infrastructure, tssc-gitops, tssc-acs
```

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced a command to inspect deployment dependency topology before installation.
  * Added a subcommand to display the dependency graph in a tabular format.
  * Added a disclaimer message for experimental deployments.

* **Enhancements**
  * Improved and expanded documentation for dependency resolution, topology, and configuration steps.
  * Helm charts now include annotations specifying product names and dependencies for clearer deployment order.
  * Product and dependency configuration structure was streamlined for clarity and consistency.

* **Bug Fixes**
  * Improved YAML editing in deployment and integration scripts for more robust configuration changes.

* **Refactor**
  * Reworked internal handling of products and dependencies, removing legacy dependency sections and updating variable naming conventions.
  * Simplified and unified chart and configuration management logic.
  * Removed intermediate chart filesystem abstraction for hooks and installer, streamlining Helm client usage.

* **Tests**
  * Added comprehensive tests for dependency resolution, topology management, and chart collection logic.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Roming22 commented on docs/topology.md:
```diff
@@ -0,0 +1,53 @@
+
+To enable dynamic dependency resolution the following annotations will be added to the `Chart.yaml` files of TSSC Helm charts. These annotations guide the installer in building the correct deployment order, and the namespaces is determined by the installer [configuration](../README.md#-installerproducts-).
+
+Product associated Helm charts will use the namespace determined in the configuration, while the other charts will use the same namespace as the installer (`.tsssc.namspace`).
```
.tsssc.namspace -> .tssc.namespace

Roming22 commented on docs/topology.md:
```diff
@@ -0,0 +1,53 @@
+
+### `tssc.redhat-appstudio.github.com/product-name`
+
+- **Purpose**: This **optional** annotation identifies the product name that a Helm chart is associated with. It links a specific Helm chart to a product defined in `config.yaml` (e.g., `.tssc.products[{name: "penShift GitOps"}]`). A single Helm chart should be associated with each product for consistency.
```
penShift GitOps -> OpenShift GitOps

Roming22 commented on installer/charts/tssc-acs/Chart.yaml:
```diff
@@ -5,3 +5,6 @@ description: TSSC Advanced Cluster Security
 appVersion: "4.8"
+annotations:
+  tssc.redhat-appstudio.github.com/product-name: Advanced Cluster Security
+  tssc.redhat-appstudio.github.com/depends-on: tssc-openshift, tssc-subscriptions
```
Isn't there a dependency between tssc-openshift and tssc-subscriptions? In which case only tssc-subscriptions is needed.

otaviof replied:
The idea is to declare all dependencies a Helm chart has, in order. This way we can enable and disable products keeping the sequence of dependencies consistent.

In a future interaction we could improve this design, or test if it already works this way, linking dependency to dependency. Please consider how the Resolver, new actor added in this PR, works for more details.

Roming22 replied:
Good enough for a first step.

Roming22 commented on installer/charts/tssc-dh/Chart.yaml:
```diff
@@ -4,3 +4,6 @@ description: Red Hat Developer Hub
 appVersion: "1.6"
+annotations:
+  tssc.redhat-appstudio.github.com/product-name: Developer Hub
+  tssc.redhat-appstudio.github.com/depends-on: tssc-openshift, tssc-subscriptions, tssc-infrastructure, tssc-pipelines, tssc-gitops, tssc-tas, tssc-tpa, tssc-app-namespaces
```
Since tssc-pipelines depends on tssc-subscriptions, it should not be necessary to declare it.

otaviof replied:
The same idea is expressed on my previous comment, btw: https://github.com/redhat-appstudio/rhtap-cli/pull/1090#discussion_r2257293992

Roming22 commented on pkg/config/product.go:
```diff
@@ -2,29 +2,21 @@ package config
-	AdvancedClusterSecurity = "advancedClusterSecurity"
-	// Quay Red Hat Quay (RHDH).
-	Quay = "quay"
+	DeveloperHub = "Developer Hub"
```
I wish we could find a way to get rid of these, as they are getting in the way of this being a framework.

otaviof replied:
It's work in progress. This PR takes us a long way towards this direction, and more will take place soon.

Roming22 replied:
Alright, to be done later then.

Roming22 commented on pkg/resolver/collection.go:
```diff
@@ -0,0 +1,125 @@
+}
+
+// DependsOn returns the list of chart names this chart depends on.
+func (c *Collection) DependsOn(hc *chart.Chart) []string {
```
As alluded in previous comments, I would expect this method to be recursive. If a depends on b, and b depends on c, I should not have to declare that a depends on c.

otaviof replied:
It's recursive, please consider the Resolver entrypoint for reference. We build the topology centred on the products enabled.

Roming22 commented on pkg/resolver/resolver.go:
```diff
@@ -0,0 +1,154 @@
+	"helm.sh/helm/v3/pkg/chart"
+)
+
+// Resolver represents theh actor that resolves dependencies between charts.
```
theh -> the

otaviof replied:
Indeed.

Roming22 commented on pkg/resolver/resolver.go:
```diff
@@ -0,0 +1,154 @@
+	// Ensure the chart is not visited again, to prevent circular dependencies.
+	chartName := hc.Name()
+	if visited[chartName] {
+		return fmt.Errorf("%w: a %q dependency is depedent itself. ",
```
"%w: a %q dependency is depedent itself. " -> "%w: product '%q' has a dependency on itself."

otaviof replied:
```go
return fmt.Errorf("%w: a %q dependency requires %q", ErrCircularDependency, chartName, chartName)
```

Roming22 commented on pkg/resolver/resolver.go:
```diff
@@ -0,0 +1,154 @@
+	return nil
+}
+
+// resolveDependencies final inspection of the Helm chrts in the collection to
```
chrts -> charts

Roming22 commented on README.md:
```diff
@@ -143,6 +137,8 @@ This is currently mainly a placeholder for future configuration settings that wo
 
 ### `{{ .Installer.Products.* }}`
 
+The product name, `.name` attribute, is converted sanitized to work as a template variable. For example, if the product name is `Developer Hub` it will be converted to `Developer_Hub`.
```
converted sanitized -> sanitized

otaviof replied:
Thanks!

otaviof commented:
/retest

otaviof commented:
/retest

otaviof commented:
/retest

otaviof commented:
/retest

Roming22 commented on docs/topology.md:
```diff
@@ -0,0 +1,53 @@
+
+To enable dynamic dependency resolution the following annotations will be added to the `Chart.yaml` files of TSSC Helm charts. These annotations guide the installer in building the correct deployment order, and the namespaces is determined by the installer [configuration](../README.md#-installerproducts-).
+
+Product associated Helm charts will use the namespace determined in the configuration, while the other charts will use the same namespace as the installer (`.tsssc.namespace`).
```
typo still present

otaviof replied:
Just fixed.

Roming22 reviewed (commented):
Some comments haven't been processed.

Roming22 commented on pkg/config/product.go:
```diff
@@ -35,17 +27,38 @@ type ProductSpec struct {
+
+	// Ensure the name doesn't start with a digit by prefixing it with an
+	// underscore if it does.
+	if len(key) > 0 && '0' <= key[0] && key[0] <= '9' {
```
Do we actually consider "" (empty string) to be a valid product name? Shouldn't it be an error?

otaviof replied:
Empty string means it's not a Helm chart product. And, every product must have a associated Helm chart, so it will error out earlier if the configuration doesn't find a Helm chart for the product.

otaviof commented:
/retest

otaviof commented:
/hold CI jobs are too quick to be true 😅

otaviof committed:
feat: Dynamic Dependency Resolver

Removing the hardcoded dependencies from the configuration and
introducing a dynamic dependency resolver that utilizes Helm chart
annotations to determine the topology.

The installer configuration has been modified to use a slice of
products instead of a map. In Golang, slices have a static order, while
maps are unordered. This change enables a predictable Helm chart
topology and removes the hardcoded product names from the
configuration.

Products have a clear description on top of each entry, the LLM will
take advantage of this information to provide clear use instructions
whether to enable or not the product.

otaviof committed:
feat: Chart Metadata

Annotating associated product names and dependencies to replace the
previously hardcoded list.

otaviof committed:
feat: Topology Subcommand

Shows the Helm chart dependency topology, as well metadata about the
charts, like inter chart dependency, namespace and associated product
name.

otaviof committed:
chore: Update Deploy Script

To comply with the new configuration schema.

otaviof committed:
feat: Topology Documentation

Updating the project documentation to include the new items.

xinredhat commented:
/retest
